lib/utils/patterns.ts

```diff
@@ -0,0 +1,228 @@
+import ResinSdk = require('resin-sdk');
+import chalk from 'chalk';
+import validation = require('./validation');
+import messages = require('./messages');
+
+const resin = ResinSdk.fromSharedOptions();
+
+export function authenticate(options: {}) {
+	form.run([{
+		message: 'Email:',
+		name: 'email',
+		type: 'input',
+		validate: validation.validateEmail
+	}, {
+		message: 'Password:',
+		name: 'password',
+		type: 'password'
+	}], { override: options })
+	.then(resin.auth.login)
+	.then(resin.auth.twoFactor.isPassed)
+	.then((isTwoFactorAuthPassed: boolean) => {
+		if (isTwoFactorAuthPassed) { return; }
+
+		return form.ask({
+			message: 'Two factor auth challenge:',
+			name: 'code',
+			type: 'input'}).then(resin.auth.twoFactor.challenge)
+		.catch((error: any) =>
+			resin.auth.logout().then(function() {
+				if ((error.name === 'ResinRequestError') && (error.statusCode === 401)) {
+					throw new Error('Invalid two factor authentication code');
+				}
+				throw error;
+			})
+		);
+	})
+};
+
+export function askLoginType() {
+	return form.ask({
+		message: 'How would you like to login?',
+		name: 'loginType',
+		type: 'list',
+		choices: [{
+			name: 'Web authorization (recommended)',
+			value: 'web'
+		}, {
+			name: 'Credentials',
+			value: 'credentials'
+		}, {
+			name: 'Authentication token',
+			value: 'token'
+		}, {
+			name: 'I don\'t have a Resin account!',
+			value: 'register'
+		}]
+	})
+}
+
+export function selectDeviceType() {
+	return resin.models.device.getSupportedDeviceTypes()
+	.then(deviceTypes => {
+		return form.ask({
+			message: 'Device Type',
+			type: 'list',
+			choices: deviceTypes
+		})
+	});
+}
+
+export function confirm(yesOption: string, message: string, yesMessage: string) {
+	return Promise.try(function () {
+		if (yesOption) {
+			if (yesMessage) { console.log(yesMessage); }
+			return true;
+		}
+
+		return form.ask({
+			message,
+			type: 'confirm',
+			default: false
+		});
+	}).then(function(confirmed) {
+		if (!confirmed) {
+			throw new Error('Aborted');
+		}
+	})
+}
+
+export function selectApplication(filter: (app: ResinSdk.Application) => boolean) {
+	resin.models.application.hasAny().then(function(hasAnyApplications) {
+		if (!hasAnyApplications) {
+			throw new Error('You don\'t have any applications');
+		}
+
+		return resin.models.application.getAll();
+	})
+	.filter(filter || _.constant(true))
+	.then(applications => {
+		return form.ask({
+			message: 'Select an application',
+			type: 'list',
+			choices: _.map(applications, application =>
+				({
+					name: `${application.app_name} (${application.device_type})`,
+					value: application.app_name
+				})
+		)});
+	})
+}
+
+export function selectOrCreateApplication() {
+	return resin.models.application.hasAny().then((hasAnyApplications) => {
+		if (!hasAnyApplications) return;
+
+		return resin.models.application.getAll().then((applications) => {
+			let appOptions: { name: string, value: string | null }[];
```

Comment:
NIT: This could be a `const` with the object type defined as a generic parameter of `map`.

Reply:
Fixed (the generics end up being a bit more awkward than you'd imagine, and inference can't handle it as we need to add a `null` value later, but the end result's not too bad)